crates/wasi-common/src/sys/windows/host_impl.rs

```diff
@@ -40,45 +40,70 @@ pub(crate) fn errno_from_win(error: winx::winerror::WinError) -> wasi::__wasi_er
+pub(crate) fn fdflags_from_win(
+    access: AccessMode,
+    mode: FileModeInformation,
+) -> wasi::__wasi_fdflags_t {
     let mut fdflags = 0;
-    // TODO verify this!
-    if mode.contains(AccessMode::FILE_APPEND_DATA) {
+
+    if access.contains(AccessMode::FILE_APPEND_DATA)
+        && !access.contains(AccessMode::FILE_WRITE_DATA)
+    {
         fdflags |= wasi::__WASI_FDFLAG_APPEND;
     }
-    if mode.contains(AccessMode::SYNCHRONIZE) {
-        fdflags |= wasi::__WASI_FDFLAG_DSYNC;
-        fdflags |= wasi::__WASI_FDFLAG_RSYNC;
+
+    if mode.contains(FileModeInformation::FILE_WRITE_THROUGH) {
+        // Only report __WASI_FDFLAG_SYNC, which is technically the only one of the O_?SYNC flags Windows supports.
         fdflags |= wasi::__WASI_FDFLAG_SYNC;
     }
-    // The NONBLOCK equivalent is FILE_FLAG_OVERLAPPED
-    // but it seems winapi doesn't provide a mechanism
-    // for checking whether the handle supports async IO.
-    // On the contrary, I've found some dicsussion online
-    // which suggests that on Windows all handles should
-    // generally be assumed to be opened with async support
-    // and then the program should fallback should that **not**
-    // be the case at the time of the operation.
-    // TODO: this requires further investigation
+
+    if !mode.contains(FileModeInformation::FILE_SYNCHRONOUS_IO_ALERT)
+        && !mode.contains(FileModeInformation::FILE_SYNCHRONOUS_IO_NONALERT)
+    {
+        fdflags |= wasi::__WASI_FDFLAG_NONBLOCK;
+    }
+
     fdflags
 }
 
-pub(crate) fn win_from_fdflags(fdflags: wasi::__wasi_fdflags_t) -> (AccessMode, Flags) {
-    let mut access_mode = AccessMode::empty();
-    let mut flags = Flags::empty();
+pub(crate) fn win_from_fdflags(
+    fdflags: wasi::__wasi_fdflags_t,
+    read: bool,
+    write: bool,
+) -> (AccessMode, Flags) {
+    let mut access_mode = AccessMode::READ_CONTROL;
 
-    // TODO verify this!
-    if fdflags & wasi::__WASI_FDFLAG_NONBLOCK != 0 {
-        flags.insert(Flags::FILE_FLAG_OVERLAPPED);
+    if read {
+        access_mode.insert(AccessMode::GENERIC_READ);
     }
+
+    if write {
+        access_mode.insert(AccessMode::GENERIC_WRITE);
+    }
+
+    // For append, grant the handle FILE_APPEND_DATA access but *not* FILE_WRITE_DATA.
+    // This makes the handle "append only".
     if fdflags & wasi::__WASI_FDFLAG_APPEND != 0 {
         access_mode.insert(AccessMode::FILE_APPEND_DATA);
+        access_mode.remove(AccessMode::FILE_WRITE_DATA);
     }
+
+    // Enable backup semantics so directories can be opened as files
+    let mut flags = Flags::FILE_FLAG_BACKUP_SEMANTICS;
+
+    // TODO: setting this flag is the correct thing to do on Windows.
+    // However, the read *must* be overlapped (i.e. asynchronous) or the underlying system calls
+    // may report operations as completed when they have not.
+    if fdflags & wasi::__WASI_FDFLAG_NONBLOCK != 0 {
+        flags.insert(Flags::FILE_FLAG_OVERLAPPED);
+    }
```

Comment:
Just a question, since there is a TODO here, do you have an idea for handling this later?

Reply:
This question also ties directly into how I think `__wasi_poll_oneoff` should be implemented for Windows 😄.

Asynchronous I/O on Windows has a model that is very difficult to shoehorn into POSIX's `EAGAIN`, `poll`, and retry model:

* You must pass in an `OVERLAPPED` structure for `ReadFile`/`WriteFile`, so no more relying on Rust's stdlib.
* The Windows model is to perform the I/O operation *first*, check for `ERROR_IO_PENDING`, wait on the event you specified to be signaled, and then determine how many bytes were transferred by calling `GetOverlappedResult`.  Unlike POSIX, you don't try the I/O operation again as Windows will complete the operation to the *buffer that was originally passed when invoking the operation*. **This is the most problematic part.**
* Supporting overlapped reads and writes on the same file handle means maintaining multiple `OVERLAPPED` structures and events, synchronizing a file pointer, and tracking what operations are taking place.
* You have to take into account multiple threads performing overlapped I/O on the same OS file handle; there should not be multiple threads waiting on the *same* overlapped event handle otherwise you might signal the wrong I/O operation.

I've spent the better part of this morning thinking about how this will work and have only come up with solutions that have very noticeable downsides.

One solution is to simply disallow overlapping reads and writes entirely.  It would make it an error to initiate another I/O operation while one is pending (e.g. `__wasi_fd_read` returns `EAGAIN` and a call to `__wasi_fd_write` while the previous read operation is pending either also immediately returns `EAGAIN` or some other error).

Additionally, since Windows will complete a write from the original buffer given on the first call that returns `EAGAIN`, there's no way to model the following POSIX behavior: call `write` that returns `EAGAIN` and then later call `write` again but with different buffer content.  The best we can do on Windows given the no-retry model is to detect this and treat as an error.  For performance, we should probably limit the check to different sizes (*ugh!*).

If we're fine with these limitations, I think this is how we might implement non-blocking I/O for WASI on Windows:

* If a file is opened `__WASI_FDFLAG_NONBLOCK`, then we allocate a structure consisting of:

  * An `OVERLAPPED` structure that will be passed to the I/O functions.
  * The Windows manual-reset event to use to signal the completion of the I/O operation.  This will be used to block `__wasi_poll_oneoff`, as needed.
  * A `Vec<u8>` that will be resized based on the call to `__wasi_fd_read` and `__wasi_fd_write` so that we can model the POSIX behavior of `EAGAIN` and retrying the I/O operation.  This will be the buffer that the OS will complete the I/O operation to.
  * A simple enum state machine consisting of `NoOperation`, `PendingRead`, `PendingWrite`, `CompletedRead`, and `CompletedWrite`.  The pending states will store how much data was requested and the completed states will store the number of bytes transferred.

* The descriptor implementation will need to maintain its own "file pointer" since Windows doesn't use a file pointer for asynchronous file handles.  This also means that this needs to be accounted for in `__wasi_fd_seek` and `__wasi_fd_tell`.

* The `__wasi_fd_read` and `__wasi_fd_write` functions would need to check to see if the file descriptor is non-blocking and perform the following:

  * If the state machine is `NoOperation`, resize the internal buffer and call `ReadFile`/`WriteFile` with the internal `OVERLAPPED` structure.  If the operation errors with `ERROR_IO_PENDING`, transition to a pending state and return `EAGAIN`.
  * If there is a pending I/O operation, poll for the current status with a non-blocking `GetOverlappedResult` call.  If the status remains pending (`PendingRead` or `PendingWrite`) immediately return `EAGAIN`.  If it is now completed, reset the event and treat as if the status was originally completed (see next bullet item).
  * If the previous I/O operation has completed, check to see if the current I/O request can be satisfied by the result:

    * If the current operation is a read and the status is `CompletedRead`, check to see that the requested range of the file (current offset + size) overlaps with what was read into the buffer.  If it does, either return an error if the previous I/O operation failed or copy the overlapping portion out and return the number of overlapping bytes read; transition to the `NoOperation` state.  If it does not overlap at all (this may occur if the file pointer moved between read calls, for example), discard the previous I/O result and behave as if the state was `NoOperation` originally (see above).
    * If the current operation is a write and the status is `CompletedWrite`, detect a mismatched write and error (`EINVAL` perhaps).  Otherwise, transition to the `NoOperation` state and return the number of bytes written or an error if the previous I/O operation failed.
    * If the current operation is a read and the status is `CompletedWrite` or the current operation is a write and the status is `CompletedRead`, return an error (no overlapping I/O operations are permitted).

* The `__wasi_poll_oneoff` implementation would:

  * Check if the file is blocking or if a non-blocking file has a status of `NoOperation`.  It should probably signal ready with a read/write byte count of 1, perhaps?
  * If the non-blocking file has a status of `PendingRead` or `PendingWrite`, wait on the event handle to signal (paying attention to the fact that `WaitForMultipleObjects` should not have duplicates in the array, so special case a request for both a read and write event on the same file descriptor).  For a signaled wait handle, reset the event handle and then transition the state to `CompletedRead` or `CompletedWrite`, storing the number of bytes transferred.  If the requested event type (read or write) matches the completed I/O operation, return the number of bytes transferred, otherwise 0 maybe?

I know this comment is crazy long and probably belongs as the basis for a design document, but we can use this discussion as basis for a proper document.

Reply:
I do wonder if we should fix the implementation in this PR reject `__WASI_FDFLAG_NONBLOCK` as `ENOTSUP` until we can properly implement non-blocking I/O.

Reply:
Wow, you have indeed thought about this for a bit ;-) That sounds super interesting. I'm wondering though if we shouldn't postpone the nonblocking ops until we draft out with @pchickey the non-blocking version of the syscalls. Perhaps then would be the best time to overhaul Windows impl to account for nonoverlap? I'm not sure myself, just thinking out loud.

Anyhow, I think that currently returing an `ENOTSUP` in case the user specifies `__WASI_FDFLAG_NONBLOCK` is a very good idea indeed! Would you mind adding that in? :-)

Reply:
I'll make `__WASI_FDFLAG_NONBLOCK` unsupported until we figure out the story for non-blocking on Windows.

Reply:
SGTM!

Reply:
Actually, this is not a problem for `fd_write` and `fd_read`, I think.
On Linux, `write` and `read` on files are inherently synchronous (`O_NONBLOCK` has no effect on regular files) and [if one desires to do asynchronous I/O on files on Linux, a separate interface called `aio` has to be used]. Similarly, `epoll` doesn't support non-blocking reads for regular files.

All in all, I think we can, at least for now, safely ignore non-blocking I/O on files.

Reply:
I think we would need to codify it in WASI that `O_NONBLOCK` is not a valid argument for files, then, even if Windows does support asynchronous file I/O.

Regarding `poll_oneoff` what is the expected behavior for descriptors to files then?  The Unix implementation is calling POSIX poll, which makes no distinctions between file, pipe, or socket in the documentation.  If file operations are always synchronous, are they always "ready"?

Reply:
>Regarding poll_oneoff what is the expected behavior for descriptors to files then? The Unix implementation is calling POSIX poll, which makes no distinctions between file, pipe, or socket in the documentation. If file operations are always synchronous, are they always "ready"?

I think so. [select(3)](https://linux.die.net/man/3/select) says that:
> File descriptors associated with regular files shall always select true for ready to read, ready to write, and error conditions.

The same for [`poll(3)`](https://linux.die.net/man/3/poll):
> Regular files shall always poll TRUE for reading and writing.

(actually, on Linux, both `select` and `poll` are implemented using the same underlying function, `do_pollfd`)

Reply:
Ah I missed that line when I was scanning for "file" (hard to do when everything is a "file descriptor), thanks.

Ok, I'm on board with:

* Updating WASI to mention that `O_NONBLOCK` has no effect on files.
* Update this PR to ignore `O_NONBLOCK` and never set `FILE_FLAG_OVERLAPPED`.
* Table further discussion about asynchronous file I/O until we have a proper asynchronous I/O API in WASI for considering overlapped I/O (would probably favor I/O completion ports anyway).

Reply:
Agreed! Btw. there's also a relevant paragraph in `open(2)`.

> Note that this flag has no effect for regular files and block devices; that is, I/O operations will (briefly) block when device activity is required, regardless of whether
 `O_NONBLOCK` is set.  Since `O_NONBLOCK` semantics might eventually be implemented, applications should not depend upon blocking behavior when specifying this flag for regular files and block devices.

Reply:
Resolved by ignoring the `__WASI_FDFLAG_NONBLOCK` on Windows (with comments explaining why).